### PR TITLE
Fix Pango Rendering

### LINKED ISSUE #144: Markup code showing in the Instructions field of the Recipe Card
https://github.com/kirienko/gourmet/blob/4e601e87c08cdf347ce0a087e495d9d5f2807634/gourmet/reccard.py#L550
It looks like the function in /gourmet/reccard.py starting at line 546 inserts line 550 into the Instructions field of the Recipe Card in order to make some of the time units look like links, but the markup is not working right.  Or maybe it is another function?  Is putting blue underlines under hours a desired feature?

Screenshot from 2020-07-14 10-32-13

### PR DESCRIPTION
**Pull request by cydanil**
This is a big PR, which was broken down already in previous PRs (#138, #137, #135, #89).
It aims to close #144 , where the markup text is shown, but not rendered.

The text is now rendered, and links work:
pango

Where it used to be:
image

There is a trick, though.
A lot of dead and broken code was removed: I replaced the home-grown html parser found in `PangoTextBuffer`, with the `TextBuffer.insert_markup`, which works great with HTML (the format used within the database).

Because of this, Pango tags, which mark where in the text do formatting start and stop, are anonymous: they are automatically generated, and not named.

Consequently, this makes the serialisation back to HTML more difficult. I opted to remove our home-grown serialiser with the standard library `html.HTMLParser`, by first serialising to Pango markup, then convert that to HTML.

And it works!

However... links are not yet handled properly. The problem is two-fold:
 - On one hand, links within Gourmet are only time-links (text that has been detected to contain time, highlighted to launch the timer). As there is not target links, serialising this back is a question of stripping text that is both coloured and underlined.
 - On the other hand, the `LinkedTextView` aims to be more universal, and accept any hyperlinks. That is difficult, because we then must keep track of them, (which `LinkedTextView` does), and replace the seralised underline and colour with the link target, when applicable.

I have not yet resolved the link issue. I've been thinking about it, and cannot come up with an elegant solution.
I think I'm not far, though, and thus I bring you this PR, which can already be reviewed, as the changes now will be lesser: 
 - [x] fix said link serialisation;
 - [x] extend the serialisation tests in `tests/test_pango_html.py`

I apologise for dropping such a FAT PR, but the code is quite entangled, and touches a few core parts of the GUI at once.

Thanks!
Cyril

### CODE REVIEW

File: gourmet/gtk_extras/pango_html.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+        super().feed(text)
+        return self.markup_text
+
+    def handle_starttag(self, tag: str, attrs: List[str]) -> None:
+        if tag == "apply_tag":
+            id_ = dict(attrs).get('id')
+            tags = self.tags.get(id_)
+            if tags is not None:
+                self.current_closing_tags = tags[1]
+                for t in tags[0]:
```
Comment (cydanil) [inline gourmet/gtk_extras/pango_html.py]: No need to iterate, as it's a string, it should be:
```python
opening, closing = tags
self.markup_text += opening
self.current_closing_tags = closing
```

Ditto when closing

**Comment by cydanil** [comment]
Links are now handled correctly, but Gourmet time links should be cleaned prior to serialisation.

Time links are a shortcut within Gourmet, and the markup does not exist prior to being processed by the `timeScanner`, so on seralisation it should be removed.

File: gourmet/gtk_extras/TextBufferMarkup.py [hunk 2]
```diff
@@ -15,337 +15,123 @@
...
+        if normal_button is not None:
+            normal_button.connect('clicked', self.remove_all_tags)
         self.tag_widgets = {}
         self.internal_toggle = False
-        self.insert_ = self.get_insert()
-        self.connect('mark-set',self._mark_set_cb)
-        self.connect('changed',self._changed_cb)
-        for w,tup in toggle_widget_alist:
-            self.setup_widget(w,*tup)
+        self.insert_: Gtk.TextMark = self.get_insert()
```
Comment (cydanil) [inline gourmet/gtk_extras/TextBufferMarkup.py]: This could be `self._insert`: it's not used outside the class

Comment (maweki) [reply gourmet/gtk_extras/TextBufferMarkup.py]: That would be better to decrease the visible interface

File: gourmet/gtk_extras/LinkedTextView.py [hunk 3]
```diff
@@ -97,7 +120,7 @@ def event_after(self,
             try:
                 start, end = buffer.get_selection_bounds()
             except ValueError:
-                # If there is nothing selected, None is return
+                # If there is nothing selected, None is returned
```
Comment (maweki) [inline gourmet/gtk_extras/LinkedTextView.py]: Can we actually return `None` then? Also, it doesn't look like `None` is returned. I only see `return False` and only few code paths that do not have a `return`.

Comment (cydanil) [reply gourmet/gtk_extras/LinkedTextView.py]: We're not returning `None`, it's bad phrasing in the comment:

`buffer.get_selection_bounds()` may return `None` if there is no selection, (as opposed to a tuple of `TextIter`.

This `try/except/else` statement could be rewritten as something like:
```python
            # Do not follow links if the user has selected something
            selection = buffer.get_selection_bounds()
            if selection is not None:
                start, end = selection
                if start.get_offset() != end.get_offset():
                    # ie. not a real selection, but the position of the cursor
                    # in widget
                    return False
```

This function basically does a check that we're not doing text selection, (which is triggered by the same events [button release].)

Now that I think of it, the selection check could be done before:

```python
    def event_after(self,
                    text_view: 'LinkedTextView',
                    event: Gdk.Event) -> bool:
        """Handle mouse clicks on time links."""
        _, button = event.get_button()

        # Check for selection
        selecting = False

        buffer = text_view.get_buffer()
        selection = buffer.get_selection_bounds()
        if selection is not None:
            if selection[0].get_offset() != selection[1].get_offset():
                selecting = True

        # Check for a left mouse click (as set by the system, not hardware).
        if event.type == Gdk.EventType.BUTTON_RELEASE and button == 1 and not selecting:
            x, y = text_view.window_to_buffer_coords(Gtk.TextWindowType.WIDGET,
                int(event.x), int(event.y))
            _, itr = text_view.get_iter_at_location(x, y)
            self.follow_if_link(text_view, itr)

        return False  # Don't process the event further
```

Comment (maweki) [reply gourmet/gtk_extras/LinkedTextView.py]: I always dislike the mental parsing when an `if` is just to choose between values. Why not the following?

`selecting = bool(selection is not None and (selection[0].get_offset() != selection[1].get_offset()))`?

Comment (cydanil) [reply gourmet/gtk_extras/LinkedTextView.py]: That also works, yes!

File: gourmet/gtk_extras/TextBufferMarkup.py [hunk 4]
```diff
@@ -382,60 +167,99 @@ def _changed_cb (self, tb):
...
+        self.nb = Gtk.Button()
+        self.nb.set_label("Normal")
         self.editBox.add(self.nb)
         self.sw = Gtk.ScrolledWindow()
         self.tv = Gtk.TextView()
         self.sw.add(self.tv)
         self.ipb = InteractivePangoBuffer(
             normal_button=self.nb)
+
+        # self.ipb.set_text("""<b>This is bold</b>. <i>This is italic</i>
```
Comment (cydanil) [inline gourmet/gtk_extras/TextBufferMarkup.py]: This shouldn't be here

**Comment by cydanil** [comment]
The text can now be formatted in the editor.

A demo here: https://i.ibb.co/SXmWdWK/format-demo.gif

~~There is an issue, where the text saved with formatting is not rendered immediately. It has to do with the text not being serialised by default, and a call has to be updated somewhere upon save callback being called.~~

~~It's a bug I'd rather fix in another PR, to limit the scope here.~~

(Nevermind, I fixed it)

File: gourmet/gtk_extras/pango_html.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+            self.is_colored_and_underlined = True
+
+    def handle_data(self, data: str) -> None:
+        target = self.links.get(data)
+
+        if self.is_colored_and_underlined and target is not None:
+            # Replace the markup tags with a hyperlink target
+            data = f'<a href="{target}">{data}</a>'
+        else:
+            data = self.current_opening_tags + data + self.current_closing_tags
```
Comment (cydanil) [inline gourmet/gtk_extras/pango_html.py]: This is incorrect.
There should be a queue of tags to close, and pop from that in `handle_endtag`.

An examples is:
`1<b>23<i>45</i>67</b>89`, which incorrectly gets serialised as `1<b>23</b><i>45</i>6789`

Comment (cydanil) [reply gourmet/gtk_extras/pango_html.py]:
```python
    pango_markup = b'GTKTEXTBUFFERCONTENTS-0001\x00\x00\x01i <text_view_markup>\n <tags>\n  <tag name="italic" priority="1">\n   <attr name="style" type="PangoStyle" value="PANGO_STYLE_ITALIC" />\n  </tag>\n  <tag name="bold" priority="0">\n   <attr name="weight" type="gint" value="700" />\n  </tag>\n </tags>\n<text>ddf<apply_tag name="bold">fd<apply_tag name="italic">df</apply_tag>fd</apply_tag>dff</text>\n</text_view_markup>\n'  # noqa
    expected = 'ddf<b>fd<i>df</i>fd</b>dff'

    ret = PangoToHtml().feed(pango_markup)
    assert ret == expected
```

**Pull request merged**